File: src/test/rustdoc-js/impl-trait.js
// ignore-order

const QUERY = [
    'Aaaaaaa -> i32',
    'Aaaaaaa -> Aaaaaaa',
    'Aaaaaaa -> usize',
    '-> Aaaaaaa',
    'Aaaaaaa',
];

const EXPECTED = [
    {
        // Aaaaaaa -> i32
        'others': [
            { 'path': 'impl_trait::Ccccccc', 'name': 'eeeeeee' },
        ],
    },
    {
        // Aaaaaaa -> Aaaaaaa
        'others': [
            { 'path': 'impl_trait::Ccccccc', 'name': 'fffffff' },
        ],
    },
    {
        // Aaaaaaa -> usize
        'others': [],
    },
    {
        // -> Aaaaaaa
        'others': [
            { 'path': 'impl_trait::Ccccccc', 'name': 'fffffff' },
            { 'path': 'impl_trait::Ccccccc', 'name': 'ddddddd' },
            { 'path': 'impl_trait', 'name': 'bbbbbbb' },
        ],
    },
    {
        // Aaaaaaa
        'others': [
            { 'path': 'impl_trait', 'name': 'Aaaaaaa' },
        ],
        'in_args': [
            { 'path': 'impl_trait::Ccccccc', 'name': 'fffffff' },
            { 'path': 'impl_trait::Ccccccc', 'name': 'eeeeeee' },
        ],
        'returned': [
            { 'path': 'impl_trait::Ccccccc', 'name': 'fffffff' },
            { 'path': 'impl_trait::Ccccccc', 'name': 'ddddddd' },
            { 'path': 'impl_trait', 'name': 'bbbbbbb' },
        ],
    },
];

File: tests/rustdoc-js/impl-trait.rs
pub trait Aaaaaaa {}

pub trait Bbbbbbb {}

impl Aaaaaaa for () {}
impl Bbbbbbb for () {}

pub fn bbbbbbb() -> impl Aaaaaaa {
    ()
}

pub struct Ccccccc {}

impl Ccccccc {
    pub fn ddddddd(&self) -> impl Aaaaaaa {
        ()
    }
    pub fn eeeeeee(&self, _x: impl Aaaaaaa) -> i32 {
        0
    }
    pub fn fffffff(&self, x: impl Aaaaaaa) -> impl Aaaaaaa {
        x
    }
    pub fn ggggggg(&self) -> impl Aaaaaaa + Bbbbbbb {
        ()
    }
}
